lsp-mode.el

```diff
@@ -2941,16 +2935,16 @@ interface TextDocumentEdit {
                       (old-file-name (lsp--uri-to-path oldUri))
                       (new-file-name (lsp--uri-to-path newUri))
                       (buf (lsp--buffer-for-file old-file-name)))
+                (condition-case nil
+                    (rename-file old-file-name new-file-name (and options (gethash "override" options)))
+                  (file-missing))
```

Comment:
can you elaborate why swallowing the error here is needed?

Reply:
Because the file might not exist on disk (e.g. you open a project directory and find-file a new file), then it's just in a buffer + LSP server memory.

I understand that this is technically better than `file-exists-p` because it avoids a TOCTOU bug.

Reply:
I see: it this version I get the following error when trying to rename class that is not saved:

```
  "message": "Oct 5, 2019, 10:49:45 PM Problem with rename for file:///home/kyoncho/Sources/lsp/dap-mode/features/fixtures/test-project/src/main/java/temp/Ss.java\nResource '/test-project/src/main/java/temp/Ss.java' does not exist.\nJava Model Exception: Core Exception [code 368] Resource '/test-project/src/main/java/temp/Ss.java' does not exist.\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.hasCompileErrors(Checks.java:652)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.checkCompileErrorsInAffectedFile(Checks.java:624)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.checkCompileErrorsInAffectedFiles(Checks.java:637)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.RenameTypeProcessor.analyzeAffectedCompilationUnits(RenameTypeProcessor.java:1016)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.RenameTypeProcessor.doCheckFinalConditions(RenameTypeProcessor.java:570)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.JavaRenameProcessor.checkFinalConditions(JavaRenameProcessor.java:44)\n\tat org.eclipse.ltk.core.refactoring.participants.ProcessorBasedRefactoring.checkFinalConditions(ProcessorBasedRefactoring.java:225)\n\tat org.eclipse.ltk.core.refactoring.Refactoring.checkAllConditions(Refactoring.java:165)\n\tat org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:82)\n\tat org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)\n\tat org.eclipse.jdt.ls.core.internal.handlers.RenameHandler.rename(RenameHandler.java:89)\n\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$21(JDTLanguageServer.java:682)\n\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$45(JDTLanguageServer.java:935)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)\nCaused by: java.lang.Exception: Resource '/test-project/src/main/java/temp/Ss.java' does not exist.\n\tat org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:42)\n\tat org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:38)\n\tat org.eclipse.core.internal.resources.Resource.checkExists(Resource.java:330)\n\tat org.eclipse.core.internal.resources.Resource.checkAccessible(Resource.java:204)\n\tat org.eclipse.core.internal.resources.Resource.findMarkers(Resource.java:976)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.hasCompileErrors(Checks.java:643)\n\t... 19 more\nCaused by: org.eclipse.core.internal.resources.ResourceException(/test-project/src/main/java/temp/Ss.java)[368]: java.lang.Exception: Resource '/test-project/src/main/java/temp/Ss.java' does not exist.\n\tat org.eclipse.core.internal.resources.ResourceException.provideStackTrace(ResourceException.java:42)\n\tat org.eclipse.core.internal.resources.ResourceException.<init>(ResourceException.java:38)\n\tat org.eclipse.core.internal.resources.Resource.checkExists(Resource.java:330)\n\tat org.eclipse.core.internal.resources.Resource.checkAccessible(Resource.java:204)\n\tat org.eclipse.core.internal.resources.Resource.findMarkers(Resource.java:976)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.hasCompileErrors(Checks.java:643)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.checkCompileErrorsInAffectedFile(Checks.java:624)\n\tat org.eclipse.jdt.internal.corext.refactoring.Checks.checkCompileErrorsInAffectedFiles(Checks.java:637)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.RenameTypeProcessor.analyzeAffectedCompilationUnits(RenameTypeProcessor.java:1016)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.RenameTypeProcessor.doCheckFinalConditions(RenameTypeProcessor.java:570)\n\tat org.eclipse.jdt.ls.core.internal.corext.refactoring.rename.JavaRenameProcessor.checkFinalConditions(JavaRenameProcessor.java:44)\n\tat org.eclipse.ltk.core.refactoring.participants.ProcessorBasedRefactoring.checkFinalConditions(ProcessorBasedRefactoring.java:225)\n\tat org.eclipse.ltk.core.refactoring.Refactoring.checkAllConditions(Refactoring.java:165)\n\tat org.eclipse.ltk.core.refactoring.CheckConditionsOperation.run(CheckConditionsOperation.java:82)\n\tat org.eclipse.ltk.core.refactoring.CreateChangeOperation.run(CreateChangeOperation.java:122)\n\tat org.eclipse.jdt.ls.core.internal.handlers.RenameHandler.rename(RenameHandler.java:89)\n\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$21(JDTLanguageServer.java:682)\n\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$45(JDTLanguageServer.java:935)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)\n",
```

I think that (again VScode does not do that) the language servers do not expect that file might exist only in memory like emacs can do. (I know it is against the spec)

I know that this might be disappointing for you since you are writing a server and you are following the spec but in a lot of cases if we follow the spec it simply wont work and a lot of server maintainers just do not care about anything but the vscode support...

Reply:
> I think that (again VScode does not do that) the language servers do not expect that file might exist only in memory like emacs can do.

I don't think this is specific to Emacs, is it? Pretty much any text editor should be able to allow editing a file that doesn't exist on disk yet.

> I know that this might be disappointing for you since you are writing a server and you are following the spec but in a lot of cases if we follow the spec it simply wont work and a lot of server maintainers just do not care about anything but the vscode support...

Well, what is the problem with following the spec here? I don't think this will get in the way of any real use cases (for Java code anyway)...

Reply:
> Well, what is the problem with following the spec here? 

Sorry,  just realized that the bug I found is not related to your PR. The bug is that if you create a new file in emacs and then try to rename the class in it without saving it fails on JDT side.